includes/class-wc-subscription.php

```diff
@@ -251,31 +251,30 @@ public function __get( $key ) {
 	}
 
 	/**
-	 * Checks if the subscription has an unpaid order or renewal order (and therefore, needs payment).
+	 * Checks if the subscription needs payment.
+	 *
+	 * A subscription requires payment if it:
+	 *  - is pending or failed,
+	 *  - has an unpaid parent order, or
+	 *  - has an unpaid order or renewal order (and therefore, needs payment)
 	 *
-	 * @param string $subscription_key A subscription key of the form created by @see self::get_subscription_key()
-	 * @param int $user_id The ID of the user who owns the subscriptions. Although this parameter is optional, if you have the User ID you should pass it to improve performance.
-	 * @return bool True if the subscription has an unpaid renewal order, false if the subscription has no unpaid renewal orders.
 	 * @since 1.0.0 - Migrated from WooCommerce Subscriptions v2.0
+	 *
+	 * @return bool True if the subscription requires payment, otherwise false.
 	 */
 	public function needs_payment() {
-
 		$needs_payment = false;
+		$parent_order  = $this->get_parent();
 
-		// First check if the subscription is pending or failed or is for $0
+		// If the subscription is pending or failed and it has a total > 0, it needs payment.
 		if ( parent::needs_payment() ) {
-
 			$needs_payment = true;
-
-		// Now make sure the parent order doesn't need payment
-		} elseif ( ( $parent_order = $this->get_parent() ) && ( $parent_order->needs_payment() || $parent_order->has_status( array( 'on-hold', 'cancelled' ) ) ) ) {
-
+		} elseif ( $parent_order && ( $parent_order->needs_payment() || $parent_order->has_status( array( 'on-hold', 'cancelled' ) ) ) ) {
+			// If the subscription has an unpaid parent order, it needs payment.
 			$needs_payment = true;
-
-		// And finally, check that the latest order (switch or renewal) doesn't need payment
 		} else {
-
-			$order = $this->get_last_order( 'all', array( 'renewal', 'switch' ) );
```

Comment:
You'll notice that switch orders have been removed as part of this change too. Switch orders are similar to early renewal orders in that an unpaid switch order isn't required to be paid. 

A long time ago, the switch occurred on checkout before payment was processed and at that time it made sense that an unpaid switch order would require payment for the subscription to continue. 

We changed switch orders to process the switch after the order is paid and so, similar to early renewals are no longer strictly required, it's an optional process.